### DISCUSSION TITLE
Saspy check for active/live connection?

### DISCUSSION THREAD

**Discussion by frbelotto** [sassoftware/saspy#555] [General]
Hello guys,
Is there a way to check if there is an existing connection instead of establishing a new one? Is it possible to check if the existing connection is alive?

I just want to create a simple code to check the questions above before creating new unneeded connections.

**Comment by tomweber-sas** [comment]
Hey, yes there is. Though, as the connection form Python to SAS is asynchronous, there's the possibility that saspy doesn't know SAS has been terminated until after it tries to make a submission to it. The way to check is the SASpid attribute of the SASsession object. While the connection is valid (alive) it has the actual SAS processes pid value. When saspy sees the connection is terminated (for whatever reason), it changes that value to None. 

So is you have a SASsession names `sas` then sas.SASpid == None will tell if the session is still connected or not. That is once saspy knows the SAS is gone on the other side. This just means that some SASPy method needs to (have been) run with the terminated connection, so it can set the SASpid value to None. If SAS terminates between when the last method ran, and before you run another method, SASpid will still be the value until saspy tries to interact with that session again. Here's an example of that case where saspy doesn't know SAS went away until you try to run some method, then it sets SASpid to None:

```
tom64-7> python3
Python 3.9.12 (main, Apr  5 2022, 06:56:58)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession(cfgname='sdssas')
SAS Connection established. Subprocess id is 965835

>>> sas
Access Method         = STDIO
SAS Config name       = sdssas
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = /sastmp/SAS_workC174000EBCE8_tom64-7/
SAS Version           = 9.04.01M8D01182023
SASPy Version         = 5.2.3
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = latin1
Python Encoding value = latin_1
SAS process Pid value = 965864


>>> sas.SASpid
'965864'
>>> sas.SASpid == None
False
>>> # kill the SAS process out from under me - from another shell 'kill -9 965864'
>>> sas.SASpid == None
False
>>> sas.assigned_librefs()
[]
>>> sas.SASpid == None
True
>>> sas
Access Method         = STDIO
SAS Config name       = sdssas
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = /sastmp/SAS_workC174000EBCE8_tom64-7/
SAS Version           = 9.04.01M8D01182023
SASPy Version         = 5.2.3
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = latin1
Python Encoding value = latin_1
SAS process Pid value = None


>>>
```

**Reply by frbelotto** [reply]
Thanks,
But if I got right, by controlling the active connection by my pid, if the connection is lost from the other side, I will be blind, right?
If i try to execute any command over a lost connection, what would be the error? Is every time the same error?

**Comment by tomweber-sas** [comment]
I'm not sure I understand the first part. You're not controlling anything by the pid, that's just a variable you can check. I set it to None when I find the connection is gone; wherever in the code that happens to be. What kind of error you get from a given method, if SAS terminates, will vary depending upon what code was running. It may throw an exception, or it may handle it and print a message (and set SASpid to None). It just depends on what code is being run and how it finds out SAS isn't there. That can also vary by access method too.

You will usually see the following message in the log too. I was curious that wasn't in the code I posted above. 
```
>>> sas.assigned_librefs()
No SAS process attached. SAS process has terminated unexpectedly.
[]
>>>
```
